docs/docs/tutorial/creating-a-source-plugin/part-4/index.mdx

```diff
@@ -0,0 +1,866 @@
+
+    ```ts:title=plugin/src/source-nodes.ts
+    // Imports
+
+    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
+      // Rest of sourceNodes
+
+      if (errors) {
+        // highlight-next-line
+        reporter.panicOnBuild(errors[0].message)
+      }
+
+      // Rest of sourceNodes
+    }
+
+    // Rest of file
+    ```
+
+    <Announcement>
+
+    **Pro tip:** Try to rely on TypeScript types to check if the error you're getting back is a valid `Error` type. If not, you might need to modify the error a little bit to fit the requirements.
+
+    </Announcement>
+
+1. To see if it works, add a typo to your GraphQL request inside `fetchGraphQL`:
+
+    ```ts:title=plugin/src/source-nodes.ts
+    const { data, errors } = await fetchGraphQL<IApiResponse>(
+      `http://localhost:4000/graphql`,
+      `query FetchApi {
+        posts {
+          id
+          slug
+    // highlight-next-line
+          title2
+          image {
+            url
+            alt
+            width
+            height
+          }
+          author
+        }
+        authors {
+          id
+          name
+        }
+      }
+      `
+    )
+    ```
+
+    Restart the `develop:site` script and you should see an error in the terminal:
+
+    ```shell
+    ERROR  UNKNOWN
+
+    Cannot query field "title2" on type "Post". Did you mean "title"?
+    ```
+
+    Yes, it works 🎉 Don't forget to remove the typo again. In [Task: Define custom errors](#task-define-custom-errors) you'll learn how to further customize this error.
+
+    <Announcement>
+
+    **Pro tip:** This is the error that the GraphQL API directly returns. You can see the same error if you go to `http://localhost:4000/graphql` (the GraphQL API from `api` folder) and try the query there. Get yourself familiar with how, when, and in which form your API throws errors.
+
+    </Announcement>
+
+You're now surfacing any errors that occur during the data fetching to your users. Use the `error`, `panicOnBuild`, and `panic` methods throughout your code to display all relevant errors to your user.
+
+**Please note:** If necessary, add context to your error messages instead of only passing the error along. This is especially true if the error could be thrown in multiple places or if the original error is not clear enough.
+
+### Task: Show activities and timings
+
+Since your source plugin is retrieving information from a third-party API the process of doing that could take a while, especially if it's fetching a lot of data. In these instances you can use the `activityTimer` method to let the user know that something is happening in the background. You shouldn't add separate `activityTimer` instances (remember: Keep the output concise and to a minimum) but instead use the `setStatus` method on `activityTimer` to update its text.
```

Comment:
Suggestion to change for brevity:
```suggestion
Since your source plugin is retrieving information from a third-party API, that process could take a while, especially if it's fetching a lot of data. In these instances you can use the `activityTimer` method to let the user know that something is happening in the background. You shouldn't add separate `activityTimer` instances (remember: Keep the output concise and to a minimum) but instead use the `setStatus` method on `activityTimer` to update its text.
```